src/hazelcore/semantics/CursorInfo.re

```diff
@@ -946,6 +1049,101 @@ let ana_cursor_found_exp =
     );
   };
 
+let rec find_uses_block =
+        (x: Var.t, block: UHExp.block, steps, index): uses_list => {
+  let UHExp.Block(lines, e) = block;
+  let (uses, shadow) = find_uses_lines(x, lines, steps @ [0], index, false);
+  shadow
+    ? uses
+    : uses @ find_uses_exp(x, e, steps @ [index + List.length(lines)]);
+}
+and find_uses_lines =
+    (x: Var.t, lines: UHExp.lines, steps, index, shadow): (uses_list, bool) =>
+  switch (lines, shadow) {
+  | ([], shadow) => ([], shadow)
+  | (_, true) => ([], true)
+  | ([line, ...lines], false) =>
+    let (line_uses, shadow) = find_uses_line(x, line, steps @ [index]);
+    let (lines_uses, shadow) =
+      find_uses_lines(x, lines, steps, index + 1, shadow);
+    (line_uses @ lines_uses, shadow);
+  }
+and find_uses_line = (x: Var.t, line: UHExp.line, steps): (uses_list, bool) =>
+  switch (line) {
+  | ExpLine(e) => (find_uses_exp(x, e, steps), false)
+  | EmptyLine => ([], false)
+  | LetLine(p, _, block) => (
+      find_uses_block(x, block, steps @ [1], 0),
+      shadow_in_pat(x, p),
+    )
+  }
+and find_uses_exp = (x: Var.t, e: UHExp.t, steps): uses_list =>
+  switch (e) {
+  | EmptyHole(_)
+  | Var(_, InVarHole(_, _), _)
+  | Var(InHole(_, _), _, _)
+  | NumLit(_, _)
+  | BoolLit(_, _)
+  | ListNil(_)
+  | Lam(InHole(_, _), _, _, _)
+  | Inj(InHole(_, _), _, _)
+  | Case(InHole(_, _), _, _, _)
+  | ApPalette(_, _, _, _) => []
+  | Var(NotInHole, NotInVarHole, y) =>
+    Printf.printf("x is used\n");
+    x == y ? [steps] : [];
+  | Lam(NotInHole, p, _, block) =>
+    shadow_in_pat(x, p) ? [] : find_uses_block(x, block, steps @ [1], 0)
+  | Inj(NotInHole, _, block) => find_uses_block(x, block, steps @ [0], 0)
+  | Case(NotInHole, block, rules, _) =>
+    find_uses_block(x, block, steps @ [0], 0)
+    @ find_uses_rules(x, rules, steps, 0)
+  | Parenthesized(block) => find_uses_block(x, block, steps @ [0], 0)
+  | OpSeq(_, opseq) =>
+    // TODO: check corresponding ErrStatus in skel_t
+    let (uses, _) = find_uses_opseq(x, opseq, steps);
+    uses;
+  }
+and find_uses_rules =
+    (x: Var.t, rules: UHExp.rules, steps, prefix_length): uses_list =>
+  switch (rules) {
+  | [] => []
+  | [rule, ...rules] =>
+    let UHExp.Rule(p, block) = rule;
+    let uses =
+      shadow_in_pat(x, p)
+        ? [] : find_uses_block(x, block, steps @ [prefix_length], 0);
+    uses @ find_uses_rules(x, rules, steps, prefix_length + 1);
+  }
+and find_uses_opseq = (x: Var.t, opseq: UHExp.opseq, steps): (uses_list, int) =>
+  switch (opseq) {
+  | ExpOpExp(e1, _, e2) => (
+      find_uses_exp(x, e1, steps @ [0]) @ find_uses_exp(x, e2, steps @ [1]),
+      2,
+    )
+  | SeqOpExp(seq, _, e) =>
+    let (uses, length) = find_uses_opseq(x, seq, steps);
+    (uses @ find_uses_exp(x, e, steps @ [length]), length + 1);
+  }
+and shadow_in_pat = (x: Var.t, p: UHPat.t): bool =>
```

Comment:
(this doesn't need to be mutually defined because it doesn't call any of the other functions, see comment above)